tools/wake-format/emitter.cpp

```diff
@@ -1572,33 +1641,66 @@ wcl::doc Emitter::walk_require(ctx_t ctx, CSTElement node) {
+                          .consume_wsnlc()
+                          .space()
+                          .token(TOKEN_P_EQUALS)
+                          .consume_wsnlc()
+                          .join(rhs_fmt())
+                          .consume_wsnlc()
+                          .newline()
+                          .fmt_if(TOKEN_KW_ELSE, else_fmt);
 
   MEMO_RET(fmt()
-               .freshline()
-               .token(TOKEN_KW_REQUIRE)
-               .ws()
-               .walk(WALK_NODE)
-               .consume_wsnlc()
-               .space()
-               .token(TOKEN_P_EQUALS)
-               .consume_wsnlc()
-               .join(rhs_fmt())
+               .join(pre_body_fmt)
+               // Returns true if the body should be separated from the currently require
                .fmt_if_else(
-                   contains_req_else,
-                   fmt(),  // Do nothing if true, else re-emit author added newlines
-                   fmt().fmt_if(TOKEN_NL, fmt().next().fmt_while(TOKEN_NL, fmt().next().newline())))
-               .consume_wsnlc()
-               .newline()
-               .freshline()
-               .fmt_if(TOKEN_KW_ELSE, else_fmt)
+                   [this, node, pre_body_fmt](const wcl::doc_builder& builder, ctx_t ctx,
+                                              const CSTElement& inner,
+                                              const token_traits_map_t& traits) {
+                     // only other requires may be next to us
+                     if (inner.id() != CST_REQUIRE) {
+                       return true;
+                     }
+
+                     // header comment forces split
+                     if (count_leading_newlines(traits, inner) > 0) {
+                       return true;
+                     }
+
+                     // if the header of the this require is multiline
+                     // force a split. + 1 because we always end the
+                     // header with a newline.
+                     if (builder->newline_count() > count_allowed_newlines(traits, node) + 1) {
+                       return true;
+                     }
+
+                     // if the header of the next require is multiline
+                     // force a split. We only check the header because
+                     // the body slurps up everything remaining in scope
+                     // thus is always many lines long. + 1 because we
+                     // always end the header with a newline.
+                     CSTElement copy = inner.firstChildElement();
+                     wcl::doc fmted =
+                         fmt().join(pre_body_fmt).compose(ctx.sub(builder), copy, token_traits);
+                     if (fmted->newline_count() > count_allowed_newlines(traits, inner) + 1) {
+                       return true;
+                     }
+
+                     return false;
+                   },
+                   fmt().newline().freshline(), fmt().freshline())
```

Comment:
I love "newline" and "freshline", I think those were very hard won and this shows how nice they are!